functional/server_test.go

```diff
@@ -0,0 +1,137 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package functional
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/coreos/fleet/functional/platform"
+	"github.com/coreos/fleet/functional/util"
+)
+
+// TestReconfigureServer checks whether fleetd managed to keep its listeners
+// across reconfiguration of fleetd after receiving SIGHUP.
+func TestReconfigureServer(t *testing.T) {
+	cluster, err := platform.NewNspawnCluster("smoke")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cluster.Destroy()
+
+	m0, err := cluster.CreateMember()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = cluster.WaitForNMachines(m0, 1)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check expected state before reconfiguring fleetd
+	stdout, _ := cluster.MemberCommand(m0, "systemctl", "show", "--property=ActiveState", "fleet")
+	if strings.TrimSpace(stdout) != "ActiveState=active" {
+		t.Fatalf("Fleet unit not reported as active: %s", stdout)
+	}
+	stdout, _ = cluster.MemberCommand(m0, "systemctl", "show", "--property=Result", "fleet")
+	if strings.TrimSpace(stdout) != "Result=success" {
+		t.Fatalf("Result for fleet unit not reported as success: %s", stdout)
+	}
+
+	// NOTE: we need to sleep once here to get reliable results.
+	// Without this sleep, the entire fleetd test always ends up succeeding
+	// no matter whether SIGHUP came or not.
+	_, _ = cluster.MemberCommand(m0, "sh", "-c", `'sleep 2'`)
+
+	err = waitForFleetdSocket(cluster, m0)
+	if err != nil {
+		t.Fatalf("Failed to get a list of fleetd sockets: %v", err)
+	}
+
+	// send a SIGHUP to fleetd, and periodically checks if a message
+	// "Reloading configuration" appears in fleet's journal, up to timeout (15) seconds.
+	stdout, _ = cluster.MemberCommand(m0, "sudo", "kill", "-HUP", "$(pidof fleetd)")
+	if strings.TrimSpace(stdout) != "" {
+		t.Fatalf("Sending SIGHUP to fleetd returned: %s", stdout)
+	}
+
+	_, err = util.WaitForState(
+		func() bool {
+			// NOTE: journalctl should run just simply like "journalctl -u fleet",
+			// without being piped with grep. Doing
+			// "journalctl -u fleet | grep \"Reloading configuration\"" is racy
+			// in a subtle way, so that it sometimes fails only on semaphoreci.
+			// - dpark 20160408
+			stdout, _ := cluster.MemberCommand(m0, "journalctl -u fleet")
+			journalfleet := strings.TrimSpace(stdout)
+			if !strings.Contains(journalfleet, "Reloading configuration") {
+				t.Logf("Fleetd is not fully reconfigured, retrying... entire fleet journal:\n%v", journalfleet)
+				return false
+			}
+			return true
+		},
+	)
+	if err != nil {
+		t.Fatalf("Fleetd reconfiguration log not found: %v", err)
+	}
+
+	err = waitForFleetdSocket(cluster, m0)
+	if err != nil {
+		t.Fatalf("Failed to get a list of fleetd sockets: %v", err)
+	}
```

Comment:
> I see that you use it here now to check for fleet being up, which makes sense I guess. With this in place, do you still need the sleep above?...

Unfortunately yes, I still need the sleep. I just did another round of tests, but it's still the same.
As mentioned above, without the sleep, the whole test always succeeds. What would be the point testing this at all, if it would always give the same answer? That's where I'm puzzled, but for me it's more important to get the test working at all.